lib/auth/webauthn/register.go

```diff
@@ -255,28 +269,30 @@ func (f *RegistrationFlow) Finish(ctx context.Context, user, deviceName string,
 	// TODO(codingllama): Verify that the public key matches the allowed
 	//  credential params? It doesn't look like duo-labs/webauthn does that.
 
-	wla, err := f.Identity.GetWebauthnLocalAuth(ctx, user)
+	wla, err := f.Identity.GetWebauthnLocalAuth(ctx, req.User)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	u := newWebUser(user, wla.UserID, true /* credentialIDOnly */, nil /* devices */)
+	u := newWebUser(req.User, wla.UserID, true /* credentialIDOnly */, nil /* devices */)
 
-	sessionDataPB, err := f.Identity.GetWebauthnSessionData(ctx, user, scopeSession)
+	sessionDataPB, err := f.Identity.GetWebauthnSessionData(ctx, req.User, scopeSession)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	sessionData := sessionFromPB(sessionDataPB)
 
-	// Activate passwordless switches (resident key, user verification) if we
-	// required verification in the begin step.
-	passwordless := sessionData.UserVerification == protocol.VerificationRequired
+	requiredUserVerification := sessionData.UserVerification == protocol.VerificationRequired
+	// If requested passwordless, check that we required verification in the begin step.
+	if req.Passwordless && !requiredUserVerification {
+		return nil, trace.BadParameter("failed to finish webauthn registration: requested passwordless but device response is not passwordless capable")
```

Comment:
```suggestion
		return nil, trace.BadParameter("passwordless registration failed, requested CredentialCreation was for an MFA registration")
```

If we didn't ask for user verification in Begin(), then we can't say anything about the device capabilities now. That is one of the reasons we reject here (the other being that we didn't ask for a resident credential either).